[mini] factor out interp into a library

lewurm opened the pull request:
_EDIT: outdated PR message, have a look at commit messages for details._

It enables us to ship the interp as an add-on to the runtime.

Usage:

configure mono without the interpreter, such as:

```
$ ./autogen.sh --enable-minimal=interpreter ...
$ make; make install
```

run your favorite program with the interpreter and see it fail:

```
$ ./mono/mini/mono-sgen --interpreter ./mono/mini/basic.exe
Could not load symbol: dlsym(RTLD_DEFAULT, mono_ee_interp_version): symbol not found
Mono IL interpreter support is missing
```

despite `--enable-minimal=interpreter`, we can still build the interp
module:

```
% make -C mono/mini libmono-ee-interp-static.la
  CC       interp/libmono_ee_interp_static_la-interp.lo
  CC       interp/libmono_ee_interp_static_la-mintops.lo
  CC       interp/libmono_ee_interp_static_la-transform.lo
  CCLD     libmono-ee-interp-static.la
```

and now can be statically linked into `mono-sgen-with-interp`:

```
(cd mono/mini && gcc -I../.. -I../../mono/eglib -I../../mono/eglib -fvisibility=hidden -O0 -g -std=gnu99 -fno-strict-aliasing -fwrapv \
-DMONO_DLL_EXPORT -g -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes -Wmissing-prototypes \
-Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes \
-Wno-format-zero-length -Qunused-arguments -Wno-unused-function -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign \
-Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -Werror-implicit-function-declaration
-o mono-sgen-with-interp mono_sgen-main-sgen.o  -framework CoreFoundation -framework Foundation \
-force_load ./.libs/libmono-ee-interp-static.a \
./.libs/libmini-static.a ../../mono/metadata/.libs/libmonoruntimesgen-static.a \
../../mono/sgen/.libs/libmonosgen-static.a ../../mono/utils/.libs/libmonoutils.a \
../../mono/eglib/.libs/libeglib.a -lm -liconv -lpthread)
```

and now you can run it again:

```
./mono/mini/mono-sgen-with-interp --interpreter ./mono/mini/basic.exe
Cannot open assembly './mono/mini/basic.exe': No such file or directory.
* Assertion: should not be reached at tramp-amd64.c:1022

Stacktrace:

  at <unknown> <0xffffffff>
  at System.Collections.Generic.EqualityComparer`1<string>.CreateComparer () <0x00006>
  at System.Collections.Generic.EqualityComparer`1<string>.get_Default () <0x0000e>
  at System.Collections.Generic.Dictionary`2<string, System.LocalDataStoreSlot>..ctor (int,System.Collections.Generic.IEqualityComparer`1<string>) <0x0005e>
  at System.Collections.Generic.Dictionary`2<string, System.LocalDataStoreSlot>..ctor () <0x0000c>
  at System.LocalDataStoreMgr..ctor () <0x00026>
  at System.Runtime.Remoting.Contexts.Context..cctor () <0x00000>
  at (wrapper runtime-invoke) object.runtime_invoke_direct_void (object,intptr,intptr,intptr) <0x00034>

Native stacktrace:

        0   mono-sgen-with-interp               0x000000010e0ea333 mono_handle_native_crash + 291
        1   mono-sgen-with-interp               0x000000010e1eeff0 sigabrt_signal_handler + 160
        2   libsystem_platform.dylib            0x00007fff63fcef5a _sigtramp + 26
        3   ???                                 0x0000000117fc7648 0x0 + 4697388616
        4   libsystem_c.dylib                   0x00007fff63df9312 abort + 127
        5   mono-sgen-with-interp               0x000000010e4449a4 mono_log_write_logfile + 420
        6   mono-sgen-with-interp               0x000000010e43a90c structured_log_adapter + 60
        7   mono-sgen-with-interp               0x000000010e469d7d monoeg_g_logv + 109
        8   mono-sgen-with-interp               0x000000010e46a174 monoeg_assertion_message + 356
        9   mono-sgen-with-interp               0x000000010e1e8366 mono_arch_get_interp_to_native_trampoline + 38
        10  mono-sgen-with-interp               0x000000010df98b73 ves_pinvoke_method + 195
        11  mono-sgen-with-interp               0x000000010df830f5 interp_exec_method_full + 5429
        12  mono-sgen-with-interp               0x000000010df81bb2 interp_exec_method + 50
```

Now it crashes because it can't emit the required trampolines. For our use case
(iOS) we static link those trampolines as it's a non-JIT environment.

/cc @spouliot @rolfbjarne

lewurm commented on mono/metadata/ee.c:
```diff
@@ -0,0 +1,79 @@
+load_ee_from_executable (const char *name, const char *opts)
+{
+	char *err;
+	MonoDl *module = mono_dl_open (NULL, MONO_DL_EAGER, &err);
```
@spouliot @rolfbjarne just to make sure: is this going to work on iOS?

with the profiler we do have a similar problem, but it's solved by generating code in mtouch to initialize a profiler module. if `dlopen` etc. is discouraged, we can do something similar with the interpreter.

vargaz replied:
I don't think we need to be able to load the interpreter dynamically only statically, i.e. if the interpreter is disabled, then the embedding app should be able to call a function in libmono-interpreter.a like mono_interp_register () or something, which would register the interpreter with the runtime.

kumpera replied:
How about if  interpreter provides a function that instals itself against the runtime it was linked against?

lewurm replied:
@kumpera that's already the case, the question is
1. what's the visibility of this function
2. who should call it

@vargaz it's not really dynamically (yet). right now it just looks up the symbol in it's own process, so it must be linked in. In the future I would like to move the JIT to the same kind of interface, along with other execution engines such as `il2cpp`, and then loading it dynamically is getting more interesting.

vargaz commented on mono/metadata/Makefile.am:
```diff
@@ -162,6 +162,8 @@ common_sources = \
 	decimal-ms.c		\
 	decimal-ms.h		\
 	domain-internals.h	\
+	ee.c			\
```
Why are these files in metadata/ ?

lewurm replied:
it follows the same idea as `metadata/profiler.c`

vargaz replied:
It should be in mini/ imho, if we really need this, since its not used by metadata/.

kumpera commented on mono/metadata/ee-internals.h:
```diff
@@ -0,0 +1,58 @@
+	gpointer (*create_method_pointer) (MonoMethod *method, MonoError *error);
+	MonoObject* (*runtime_invoke) (MonoMethod *method, void *obj, void **params, MonoObject **exc, MonoError *error);
+	void (*init_delegate) (MonoDelegate *del);
+#ifndef DISABLE_REMOTING
```
Since this function pointer is optional in the structure, it must be part of the version number/string.

lewurm replied:
good point!

vargaz replied:
Or the field should not be optional.

vargaz commented:
I think the registration function should be public, and it should be called by:
- the runtime if the interpreter is enabled.
- the embedding app if its disabled.

vargaz commented on mono/mini/interp/interp.c:
```diff
@@ -5529,15 +5532,17 @@ interp_stop_single_stepping (void)
-mono_ee_interp_version ()
+static int
+interp_ee_version (void)
 {
```
This function is not really needed.

vargaz commented on mono/mini/interp/interp.c:
```diff
@@ -5529,15 +5532,17 @@ interp_stop_single_stepping (void)
 {
+	g_assert (mono_ee_api_version () == interp_ee_version ());
+
 	mono_native_tls_alloc (&thread_context_id, NULL);
```
I think the registration function should just register the interpreter with the runtime, it shouldn't do too much initialization, because we don't know when the embedding app will call it, it can call it when nothing has been initialized yet.

lewurm replied:
the embedding app has to call it before `mini_init ()`, that's for sure. I'm not sure what you mean by "too much initialization", it doesn't do a lot already.

vargaz replied:
I mean it might call some init function which depends on the runtime being initialized.

lewurm commented:
I would like to wait for https://github.com/mono/mono/pull/6429 before merging and then cleanup this PR.

vargaz commented:
#6429 might take some time to be merged.

lewurm commented:
ok, then let's get this puppy merged. I cleaned up the commits etc, please have another look @vargaz. one tiny change to before: I've added a `interp_init_done` flag.

vargaz commented:
Looks ok. We might want to rename mini_get_interp_callbacks () to get_ee_callbacks ().

lewurm commented:
no, interp is a specific instance of an ee. we could have `mini_get_jit_callbacks`, `mini_get_il2cpp_callbacks` etc. in the future

akoeplinger commented:
@monojenkins build

luhenry commented:
There are undefined references to `mono_ee_interp_init` in `mini_init`

lewurm commented:
@luhenry hmm, I thought that would be the VES (virtual execution system). But apparently VES is an alias for EE (execution engine) according to ECMA 335. *Sigh*, I was so happy with EE 🙂 

Compilation engine wouldn't be right word either imho, because it is very much about the execution (and hooks relevant for it, such as frame introspection).

Since it's not promoted to the API yet, how about we give it some time and see where we go with it? As of now, it's just an idea and only works with interp. It will change a lot once we have done the integration with the JIT.

Therzok commented on mono/mini/interp/interp.h:
```diff
@@ -28,48 +28,10 @@ struct _InterpMethodArguments {
-void mono_interp_init (void);
+/* must be called either
+ *  - by mini_init ()
+ *  - xor, before mini_init () is called (embedding scenario).
```
Was this supposed to be 'or'?

lewurm replied:
no 🙂  `mono_interp_init ()` is supposed to be called once, otherwise it will assert.

jaykrell replied:
"execution" makes more sense than "compilation", as a way to abstract interpreter, JIT, AOT -- they are executing one way or another, but they aren't necessarily compiling. Is VES/EE the larger system, like the entire runtime? Perhaps nobody has specified a finer grained break down? You know, to do that, would require stuff like talking about how the pieces are connected, you can't split up gc and jit/interp/aot w/o talking about barriers and suspend. If you smush everything together architecturally/"documentationally" then from the outside you can just say it is magic. And truly, from the "C#" point of view, that is perhaps the right level of specification. Implementers can chose to smush it all together and have one of everything, intertwined, or have plugability of JIT/interp/AOT or maybe other parts. What do the .NET Native people call things? Or the IL2CPP people?

joshpeterson replied:
I happen to like "execution engine", but we don't have a specific name for something like this from the IL2CPP side.

joshpeterson commented on mono/mini/ee.h:
```diff
@@ -0,0 +1,59 @@
+
+typedef gpointer MonoInterpFrameHandle;
+
+struct _MonoEECallbacks {
```
From the IL2CPP side, this looks like what we need to more easily integrate IL2CPP on the Mono runtime. Thanks!
